ports/fupy/Makefile

```diff
@@ -46,7 +46,12 @@ ifneq ($(DEBUG), 1)
 CFLAGS += -DNDEBUG
 endif
 
-LDFLAGS += -T litex.ld -Wl,-Map=$@.map -Wl,--cref 
+ifeq ($(COPY_TO_MAIN_RAM), 1)
+LDFLAGS += -T litex.ld -Wl,-Map=$@.map -Wl,--cref
+else
+LDFLAGS += -T litex-flash.ld -Wl,-Map=$@.map -Wl,--cref
```

Comment:
>Because I suspect we'll want to use the "flash" approach on other small systems which might have the code stored somewhere else again.
@ewenmcneill You touch upon an good point that I wish was documented outside the source code. I agree that `xip` is more general than `flash`. Right now, the only provision for XIP in LiteX is from flash; there exists support to XIP from a block RAM in the BIOS C code, but AFAIK it's not wired up to anything else in LiteX. 

To make matters more confusing, the _traditional_ way to run a program from block RAM in MiSoC/LiteX (emphasis on both) is to allocate block RAM as the `main_ram` section, and copy your main program to the block RAM allocated as `main_ram`. This method works just fine with both `serialboot` and `flashboot`, but naturally isn't XIP :). Support for XIP from block RAM would need to interfaced to the rest of LiteX. 

At present I'm skeptical of its advantages of "XIP from BRAM" other than a faster boot time; `litex-buildenv`'s output format means the firmware would still be appended (even if it's embedded in the gateware bitstream), so the size savings disappears. And changing the firmware implies recompiling the bitstream.

Here's a table of in-principle supported configuations (N- No, NI- Not Implemented, Y- Yes):

|Firmware Location\Method|XIP|Copy-to-RAM|
|--------------------------|---|--------------------|
|Flash                              | Y |            N               |
|Block RAM                     | NI |              Y            |
|Flash                              | N|             Y               |

Reply:
@cr1901 I'm unclear what the first/third choices are meant to be (they're both "Flash" :-) ).   Maybe you just meant that the "stored in flash" case can be either execute in place, or copy to RAM?

From your discussion with @mithro it sounds like maybe there are three options for linking here: a "copy to RAM (from somewhere)", a "execute in place from flash" and an "execute in place from block RAM" (being used to emulate  ROM).  Maybe these do need different link scripts?  In which case I'd be okay with `litex-ctr.ld` and `litex-flash.ld` if the `-flash.ld` one is in fact specific to the "execute in place from flash" case.  (If it's not obvious, maybe just call them `litex-ctr.ld` and `litex-xip.ld` at this point, and we can always rename them later when we add a third one.)

Ewen

Reply:
@ewenmcneill I updated the table :P. 

>a "copy to RAM (from somewhere)", a "execute in place from flash" and an "execute in place from block RAM" **(being used to emulate ROM)**.

Correct, emphasis mine.

"Execute in place from block RAM" and "Execute in place from flash" could probably share a linker script; the only thing in practice that would differ between the two scripts are the [output section names](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/4/html/Using_ld_the_GNU_Linker/sections.html), which are generated by LiteX into `include/generated/regions.ld`.

Right now, the output section that's used by all the XIP firmware<sup>1</sup> is called `user_flash`. Just like `COPY_TO_MAIN_RAM` is tied to the `main_ram` section<sup>2</sup>, probably nothing prevents us from creating a generic output section  to accommodate both block RAM and flash-based XIP. 

## Footnotes
[1] Except the BIOS, which is special. BIOS always targets the `rom` output section. It is technically always XIP regardless of whether it's running from block RAM (which is the default) or flash (used for constrained targets like mimasv2 and tinyFPGA). But it is hardcoded by convention to use the `rom` output section and probably _should_ remain cleanly separated from the main program.
[2] `COPY_TO_MAIN_RAM` mode is exclusively gated based on whether the `main_ram` section is defined in your SoC or not.

Reply:
@cr1901 Thanks for the details.  Sounds like we're probably best off assuming separate linker scripts for the "flash" and "BRAM emulating ROM" cases for now, so I understand the choice of `litex-flash.ld` here, which was where we started.  If we come up with a more elaborate system where they can share a linker script later it'll be easy enough to rename the `.ld` file and update the `Makefile` later.

Ewen

Reply:
There is at present no way to distinguish running from BRAM-as-ROM versus running from flash. And I think adding such detection will even further complicate `crt0` and `litex-*.ld` detection. In fact, the more I think about, the more I think this shouldn't be LiteX's responsibility in the first place to automagically determine which linker-script should be run.

But right now, I can't think of a better solution. My opinion is: "if BRAM-as-ROM automagic is supported in the future, make sure it uses the same linker script as `litex-flash.ld` (and rename `litex-flash.ld` of course). If we elect for an alternate solution for choosing a linker script, we can do as we please wrt naming the linker script :).